gr-qtgui/lib/matrix_display.cc

```diff
@@ -0,0 +1,314 @@
+    d_menu->addMenu(d_color_map_menu);
+}
+void matrix_display::set_contour(bool contour)
+{
+    if (contour) {
+        for (double level = 0.5; level < 10.0; level += 1.0)
+            d_contour_levels += level;
+        d_spectrogram->setContourLevels(d_contour_levels);
+        d_spectrogram->setDisplayMode(QwtPlotSpectrogram::DisplayMode::ContourMode, true);
+    } else
+        d_spectrogram->setDisplayMode(QwtPlotSpectrogram::DisplayMode::ImageMode, true);
+}
+
+void matrix_display::set_color_map(const std::string& color_map)
+{
+    if (color_map == "rgb") {
+        d_colorMap = new QwtLinearColorMap(Qt::darkCyan, Qt::red, QwtColorMap::RGB);
+        d_colorMap->addColorStop(0.1, Qt::cyan);
+        d_colorMap->addColorStop(0.6, Qt::green);
+        d_colorMap->addColorStop(0.95, Qt::yellow);
+
+        d_z_colorMap = new QwtLinearColorMap(Qt::darkCyan, Qt::red, QwtColorMap::RGB);
+        d_z_colorMap->addColorStop(0.1, Qt::cyan);
+        d_z_colorMap->addColorStop(0.6, Qt::green);
+        d_z_colorMap->addColorStop(0.95, Qt::yellow);
+
+    } else if (color_map == "indexed") {
+        d_colorMap = new QwtLinearColorMap(Qt::blue, Qt::red, QwtColorMap::Indexed);
+
+        d_z_colorMap = new QwtLinearColorMap(Qt::blue, Qt::red, QwtColorMap::Indexed);
+
+    } else if (color_map == "alpha") {
+        QwtAlphaColorMap* alpha_map = new QwtAlphaColorMap();
+        alpha_map->setColor(Qt::darkBlue);
+        d_colorMap = (QwtLinearColorMap*)alpha_map;
+        d_z_colorMap = (QwtLinearColorMap*)alpha_map;
+    }
+}
+
+void matrix_display::set_interpolation(const std::string& interpolation)
+{
+    //'BilinearInterpolation','NearestNeighbour'
+    if (interpolation == "BilinearInterpolation") {
+        d_data.setResampleMode(QwtMatrixRasterData::ResampleMode::BilinearInterpolation);
+    } else if (interpolation == "NearestNeighbour") {
+        d_data.setResampleMode(QwtMatrixRasterData::ResampleMode::NearestNeighbour);
+    }
+}
+
+void matrix_display::set_x_axis_label(const std::string& x_axis_label)
+{
+    d_x_axis = d_plot->axisWidget(QwtPlot::xBottom);
+    d_x_axis->setTitle(x_axis_label.c_str());
+    d_plot->enableAxis(QwtPlot::xBottom, true);
+}
+
+void matrix_display::set_y_axis_label(const std::string& y_axis_label)
+{
+    d_y_axis = d_plot->axisWidget(QwtPlot::yLeft);
+    d_y_axis->setTitle(y_axis_label.c_str());
+    d_plot->enableAxis(QwtPlot::yLeft, true);
+}
+
+void matrix_display::set_z_axis_label(const std::string& z_axis_label)
+{
+    d_z_axis = d_plot->axisWidget(QwtPlot::yRight);
+    d_z_axis->setTitle(z_axis_label.c_str());
+    d_z_axis->setColorBarEnabled(true);
+    d_z_axis->setColorMap(QwtInterval(d_z_min, d_z_max), d_z_colorMap);
+    d_plot->setAxisScale(QwtPlot::yRight, d_z_min, d_z_max);
+    d_plot->enableAxis(QwtPlot::yRight, true);
+}
+
+
+void matrix_display::save_figure()
+{
+    QPixmap qpix = this->grab();
+
+    QString types = QString(tr("JPEG file (*.jpg);;Portable Network Graphics file "
```

Comment:
Would it be possible to have PNG be the first choice here? JPEG is lossy and makes ugly artifacts at sharp edges, and isn't true to colors – all properties of an image format that we don't want for the pictures that come out of your code!

Reply:
Sure